llvm/tools/sycl-post-link/sycl-post-link.cpp

```diff
@@ -751,6 +751,63 @@ static bool removeSYCLKernelsConstRefArray(Module &M) {
   return true;
 }
 
+SmallVector<module_split::ModuleDesc, 2>
+handleESIMD(module_split::ModuleDesc &&MDesc, bool &Modified,
```

Comment:
The first parameter `MDesc` should be a parameter by value.
Otherwise, an ownership transfer doesn't happen. It happens in move constructors.

Reply:
`std::move` is just a `static_cast` to `&&`. It does no actions in run-time.
Accordingly, an ownership transfer happens in constructors if we talk about RAII idiom.

The following example demonstrates that the original instance `a` still owns the content even after you pass object to the function by rvalue .
https://ideone.com/p7wS2J

Reply:
Considering that we don't have user-provided definitions of move and copy constructor for `ModuleDesc` class, I don't see why switching from `&&` to `&` would change anything here in context of the ownership. By using rvalue references here, we imply that the original object is not needed anymore at call site and we are ok with its ownership being transferred down the call stack. 

> The following example demonstrates that the original instance a still owns the content even after you pass object to the function by rvalue .
> https://ideone.com/p7wS2J

From [class.copy.ctor.8](https://eel.is/c++draft/class.copy.ctor#8):

> If the definition of a class X does not explicitly declare a move constructor, a non-explicit one will be implicitly declared as defaulted if and only if
> - [(8.1)](https://eel.is/c++draft/class.copy.ctor#8.1) X does not have a user-declared copy constructor,
> - [(8.2)](https://eel.is/c++draft/class.copy.ctor#8.2) X does not have a user-declared copy assignment operator,
> - [(8.3)](https://eel.is/c++draft/class.copy.ctor#8.3) X does not have a user-declared move assignment operator, and
> - [(8.4)](https://eel.is/c++draft/class.copy.ctor#8.4) X does not have a user-declared destructor[.](https://eel.is/c++draft/class.copy.ctor#8.sentence-1)
>
> [Note [3](https://eel.is/c++draft/class.copy.ctor#note-3): When the move constructor is not implicitly declared or explicitly supplied, expressions that otherwise would have invoked the move constructor might instead invoke a copy constructor[.](https://eel.is/c++draft/class.copy.ctor#8.sentence-2) — end note]

That example above has a user-defined destructor and therefore compiler doesn't add any implicit move-constructor and performs fallback to copy-constructor.

Please note that even if you remove user-provided destructor, you still can't rely on the exact state of a moved object, both examples below behave in the same manner:
- https://ideone.com/BVCMMY
- https://ideone.com/x62VKO

From [cppreference](https://en.cppreference.com/w/cpp/string/basic_string/basic_string):

> Move constructor. Constructs the string with the contents of other using move semantics. `other` is left in valid, but unspecified state.

`std::string` still owning its content even after it was moved to another object is a _valid_ behavior and can't be relied upon.

Reply:
Discussed offline, by some reason I misunderstood the original comment and though that the suggestion was to use lvalue reference, while it is actually to use by-value argument with `std::move` at call site. I guess we have a number of cases where we could improve how rvalue references are used, I will follow-up on that in a separate PR.